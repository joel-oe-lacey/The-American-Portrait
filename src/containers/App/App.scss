@import '../../variables';

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

#root,
body,
html,
.App {
  height: 100%;
  width: 100%;
}