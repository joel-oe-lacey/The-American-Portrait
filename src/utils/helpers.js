export class ArtPiece {
    primaryimageurl;
    title;
    objectid;
    description;
    century;
    dated;
    culture;
    classification;
    technique;
    medium;
    colors;
    creditline;
    labeltext;

    constructor(rawArtData) {
        Object.assign(this, rawArtData);
    }
};

export const restructureArtPiece = (rawArtData) => {
    const {
        title,
        titles,
        objectid,
        description,
        provenance,
        commentary,
        labeltext,
        classification,
        creditline,
        century,
        culture,
        medium,
        videos,
        datebegin,
        dateend,
        dated,
        period,
        technique,
        colors,
        primaryimageurl,
        images,
    } = rawArtData;

    return {
        title,
        titles,
        objectid,
        description,
        provenance,
        commentary,
        labeltext,
        classification,
        creditline,
        century,
        culture,
        medium,
        videos,
        datebegin,
        dateend,
        dated,
        period,
        technique,
        colors,
        primaryimageurl,
        images
    }
}

export const bucketArtByDate = (allArt) => {
    const recompileCollection = (bucketedCollection) => {
        return Object.keys(bucketedCollection).reduce((flatColl, dateKey) => {
            flatColl = flatColl.concat(bucketedCollection[dateKey])
            return flatColl;
        }, [])
    }

    const bucketedCollection = allArt.reduce((bucketedArt, piece) => {
        const date = piece.dateend;
        switch(true) {
            case (date < 1800): 
                if (bucketedArt.toEighteenHundreds.length < 10) {
                    bucketedArt.toEighteenHundreds.push(piece)
                }
                break;
            case (date <= 1850):
                if (bucketedArt.eighteenFiftys.length < 10) {
                    bucketedArt.eighteenFiftys.push(piece)
                }
                break;
            case (date <= 1880):
                if (bucketedArt.eighteenEighty.length < 10) {
                    bucketedArt.eighteenEighty.push(piece)
                }
                break;
            case (date <= 1900):
                if (bucketedArt.toNineteenHundreds.length < 10) {
                    bucketedArt.toNineteenHundreds.push(piece)
                }
                break;
            case (date <= 1920):
                if (bucketedArt.toNineteenTwenties.length < 10) {
                    bucketedArt.toNineteenTwenties.push(piece)
                }
                break;
            case (date <= 1940):
                if (bucketedArt.toNineteenFourties.length < 10) {
                    bucketedArt.toNineteenFourties.push(piece)
                }
                break;
            case (date <= 1960):
                if (bucketedArt.toNineteenSixties.length < 10) {
                    bucketedArt.toNineteenSixties.push(piece)
                }
                break;
            case (date <= 1980):
                if (bucketedArt.toNineteenEighties.length < 10) {
                    bucketedArt.toNineteenEighties.push(piece)
                }
                break;
            case (date <= 2000):
                if (bucketedArt.toTwoThousands.length < 10) {
                    bucketedArt.toTwoThousands.push(piece)
                }
                break;
            case (date <= 2020):
                if (bucketedArt.toTwentyTwenties.length < 10) {
                    bucketedArt.toTwentyTwenties.push(piece)
                }
                break;
            default:
                break;
        }
        return bucketedArt;
    }, {
        "toEighteenHundreds": [],
        "toEighteenFiftys": [],
        "toEighteenEighty": [],
        "toNineteenHundreds": [],
        "toNineteenTwenties": [],
        "toNineteenFourties": [],
        "toNineteenSixties": [],
        "toNineteenEighties": [],
        "toTwoThousands": [],
        "toTwentyTwenties": [],
    })

    return recompileCollection(bucketedCollection);
}

