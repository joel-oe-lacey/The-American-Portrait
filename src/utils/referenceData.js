export const introText = ['People are trapped in history, and history is trapped in them" said James Baldwin in his 1953 essay "Stranger in the Village". At the time he spoke of the indivisibility of black culture from the larger collective American ethos and culture.', 
'A continuation of that allegory, I believe, also speaks to the importance of remaining in touch with our collective history. It, in similar duality, is a part of us all.', 'The American Portrait looks to reunite us with a visual piece of that history. Utilizing the Harvard Museum API, you can view a timeline of historic photos by state, ten per decade where available.','While incomplete and historically biased, these photos offer us at very least a partial window into some of the lives, events, and adolescent places that would go on to form the nation as we know it. ']

export const stateNames = ['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania', 'Puerto Rico', 'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming']

export const mockCollectionPrimary = {
    "title": "House and [blue] Car, Near Akron, Alabama",
    "objectid": 354191,
    "creditline": "Harvard Art Museums/Fogg Museum, Fund for the Acquisition of Photographs",
    "century": "20th century",
    "technique": "Dye transfer print",
    "dateend": 1978,
    "dated": "1978, printed 1991",
    "people": [
        {
        "alphasort": "Christenberry, William",
        "birthplace": "Tascaloosa, Alabama",
        "name": "William Christenberry",
        "prefix": null,
        "personid": 62113,
        "gender": "unknown",
        "role": "Artist",
        "displayorder": 1,
        "culture": "American",
        "displaydate": "born 1936",
        "deathplace": null,
        "displayname": "William Christenberry"
        }
    ],
    "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:761794"
}

export const mockCollectionSecondary = {
    "title": "Night Spot, Marion, Alabama",
    "objectid": 354192,
    "creditline": "Harvard Art Museums/Fogg Museum, Fund for the Acquisition of Photographs",
    "century": "20th century",
    "technique": "Dye transfer print",
    "dateend": 1972,
    "dated": "1972, printed 1991",
    "people": [
        {
            "alphasort": "Christenberry, William",
            "birthplace": "Tascaloosa, Alabama",
            "name": "William Christenberry",
            "prefix": null,
            "personid": 62113,
            "gender": "unknown",
            "role": "Artist",
            "displayorder": 1,
            "culture": "American",
            "displaydate": "born 1936",
            "deathplace": null,
            "displayname": "William Christenberry"
        }
    ],
    "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:761793"
}