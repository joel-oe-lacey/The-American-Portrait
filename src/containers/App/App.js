import React, { Component } from 'react';
import './App.scss';
import { apiCall } from '../../utils/fetchCalls';
import Piece from '../../components/Piece/Piece';
import Loading from '../../components/Loading/Loading';
import Timeline from '../../components/Timeline/Timeline';
import { loadRegions } from '../../actions';
import { connect } from 'react-redux';
import { restructureArtPiece, bucketArtByDate } from '../../utils/helpers';
import { loadCollection, loadSubsqCollection } from '../../actions';
import { Route, Switch } from 'react-router-dom';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      fetchComplete: false
    }
  }

  retrieveInitialCollection = async () => {
    //impliment base query here, future queries will build dynamically
    const collection = await apiCall('https://api.harvardartmuseums.org/object?apikey=b59b0050-58c4-11ea-b831-f76084e9f972&hasimage=1&place=2028429&classification=17&culture=American&size=100&sort=dateend&sortorder=desc')
    const rawCollectionResp = await collection.json();
    const structuredCollectionData = rawCollectionResp.records.reduce((restrCollection, item) => {
      if (item.primaryimageurl && item.dateend) {
        restrCollection.push(restructureArtPiece(item))
      }
      return restrCollection;
    }, [])
    this.props.loadCollectionToStore(structuredCollectionData)
    if (rawCollectionResp.info.pages > 1) {
      this.retrieveSubseqCollections(rawCollectionResp)
    } else {
      const bucketedCollection = bucketArtByDate(structuredCollectionData);
      this.props.loadCollectionToStore(bucketedCollection);
      this.setState({fetchComplete: true})
    }
  }

  retrieveSubseqCollections = async (prevCollection) => {
    const collection = await apiCall(`${prevCollection.info.next}`)
    const rawCollectionResp = await collection.json();
    const structuredCollectionData = rawCollectionResp.records.reduce((restrCollection, item) => {
      if (item.primaryimageurl && item.dateend) {
        restrCollection.push(restructureArtPiece(item))
      }
      return restrCollection;
    }, []);
    this.props.loadSubsqCollectionToStore(structuredCollectionData);
    if (rawCollectionResp.info.page !== rawCollectionResp.info.pages) {
      this.retrieveSubseqCollections(rawCollectionResp)
    } else {
      const finalCollection = this.props.collections;
      const bucketedCollection = bucketArtByDate(finalCollection);
      this.props.loadCollectionToStore(bucketedCollection);
      this.setState({ fetchComplete: true })
    }
  }

  //need to run through bucketing after collection has fired off
  //but need to account for async completion 
  componentDidMount() {
    this.retrieveInitialCollection()
  }

  render() {
    const collection = this.props.collections;
    const fetchComplete = this.state.fetchComplete;

    return (
      <div className="App">
        {!fetchComplete && <Loading />}
        <Route exact path="/" render={() => {
          return (
            fetchComplete && <Timeline />
        )}} />
        <Route exact path='/piece/:id' render={({ match }) => {
          const artPiece = collection.find(piece => piece.objectid === parseInt(match.params.id))
          return (
            <section className="App">
              {!fetchComplete && <Loading />}
              {fetchComplete && <Piece {...artPiece} />}
            </section>
          )
        }} />
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loadRegionsToStore: (regions) => { dispatch(loadRegions(regions)) },
  loadCollectionToStore: (collection) => { dispatch(loadCollection(collection)) },
  loadSubsqCollectionToStore: (collection) => { dispatch(loadSubsqCollection(collection)) }
});

export const mapStateToProps = state => ({
  collections: state.collections,
});

export default connect(mapStateToProps, mapDispatchToProps)(App)


