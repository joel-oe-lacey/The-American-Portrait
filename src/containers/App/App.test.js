import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps } from './App';
import { mockCollectionPrimary, mockCollectionSecondary } from '../../utils/referenceData';


describe('App', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('should return an array of collection objects as a prop', () => {
      const mockState = {
        region: 'Colorado',
        collections: [mockCollectionPrimary, mockCollectionSecondary],
      }
      const expected = {
        collections: [mockCollectionPrimary, mockCollectionSecondary]
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected)
    })
  })
});
