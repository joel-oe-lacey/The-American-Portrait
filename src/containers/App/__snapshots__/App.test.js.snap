// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(About)",
          "type": [Function],
        }
      }
      exact={true}
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Region)",
          "type": [Function],
        }
      }
      exact={true}
      path="/region"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Timeline)",
          "type": [Function],
        }
      }
      exact={true}
      path="/timeline"
    />
    <Route
      exact={true}
      path="/piece/:id"
      render={[Function]}
    />
    <Connect(Invalid) />
  </Switch>
</div>
`;
